aten/src/ATen/core/thread_pool.cpp

```diff
@@ -3,23 +3,131 @@
 
 namespace c10 {
 
-ThreadPool::ThreadPool() { /* intended to leave empty */ }
+ThreadPool::ThreadPool(std::size_t pool_size, int numa_node_id)
+    : threads_(pool_size),
+      running_(true),
+      complete_(true),
+      available_(pool_size),
+      total_(pool_size),
+      numa_node_id_(numa_node_id) {
+  for (std::size_t i = 0; i < pool_size; ++i) {
+    threads_[i] = std::thread(std::bind(&ThreadPool::main_loop, this, i));
+  }
+}
 
-ThreadPool::~ThreadPool() { /* intended to leave empty */ }
+ThreadPool::~ThreadPool() {
+  // Set running flag to false then notify all threads.
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    running_ = false;
+    condition_.notify_all();
+  }
 
-void ThreadPool::schedule(std::function<void(void)> task) {
-  tasks.push(task);
+  try {
+    for (auto& t : threads_) {
+      t.join();
+    }
+  } catch (const std::exception&) {
+  }
+}
+
+size_t ThreadPool::size() const {
+  return threads_.size();
+}
+
+size_t ThreadPool::numAvailable() const {
+  return available_;
+}
+
+void ThreadPool::run(const std::function<void()>& func) {
+  std::unique_lock<std::mutex> lock(mutex_);
+
+  // Set task and signal condition variable so that a worker thread will
+  // wake up and use the task.
+  tasks_.push(task_element_t(func));
+  complete_ = false;
+  condition_.notify_one();
+}
+
+void ThreadPool::waitWorkComplete() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  while (!complete_) {
+    completed_.wait(lock);
+  }
 }
 
 void ThreadPool::workOnTasksUntilCompleted(
     c10::intrusive_ptr<ivalue::Future> future) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (future->completed()) {
```

Comment:
The listeners list adds overhead to the main_loop regardless of whether it is being used. Isn't it possible to just create a condition variable and add a callback to the future that will call notify on it?

```
std::unique_lock<std::mutex> lock(future->mutex);
std::condition_variable finished;
future->addCallback([&]{
  finished.notify_all();
});
while(!future->completed()) {
  finished.wait(lock);
}
```